### DISCUSSION TITLE
Grant ALL then Revoke does not work for single column

### DISCUSSION THREAD

**Discussion by NanoCode012** [supabase/supabase#7428] [Questions] [answered]
On a new supabase project, I tried to apply column level security using the answer posted in https://github.com/supabase/supabase/discussions/3158 . 

However, I noticed that if I were to 
```sql
GRANT ALL ON person TO supabase_admin, authenticated;

REVOKE UPDATE (created_at) ON person FROM supabase_admin, authenticated; -- also tried with "created_at"
```

vs

```sql
GRANT ALL ON person TO supabase_admin, authenticated;

REVOKE UPDATE ON person FROM supabase_adminauthenticated;
```

Using the JS library in an authenticated acc (logged-in) with the ANON api key used, the former would incorrectly allow changes to the column `created_at`, whereas the latter revokes all update correctly.

```js
const { data, error } = await supabase
          .from("person")
          .update({ created_at: new Date().toISOString() });
```

I would like to prevent users from being able to update columns like `created_at`, `id`, etc. As the previous discussion has halted, I was hoping I could get some more up to date info here. A view is not so good as I do want the user to be able to modify other columns, see the `created_at` value but not modify it. I am unsure if views can do that.

Edit: I would also not prefer RPC (https://github.com/supabase/supabase/discussions/346#discussioncomment-182629). In this discussion, the fields are granted for specific columns directly instead of the above method which grants all then revokes for certain column. I was wondering if the above method could be used as it reduces any need to keep updating the grants or revoke whenever we add more columns.

Edit2: Found https://www.postgresql.org/docs/current/sql-grant.html `Granting the privilege at the table level and then revoking it for one column will not do what one might wish: the table-level grant is unaffected by a column-level operation.`. 

Edit3: As a temporary workaround while I wait for a response, https://stackoverflow.com/a/48112664 provides a method to get all column for a table in a neat way. Still a bit of a hack but could work for now.

**Comment by NanoCode012** [comment, marked as answer]
A temporary workaround based on https://stackoverflow.com/a/48112664:

Run the below while replacing `<table>` and `<privileges>`(INSERT,UPDATE...) with the correct one.
```sql
select 'GRANT <privileges>(' || array_to_string(array_agg(attname), ', ') || ') ON <table> to authenticated;' -- optionally add public,anon
  from pg_attribute
 where attrelid = '<table>'::regclass and attnum > 0;
```
Then,
```sql
-- optionally revoke all first to clear previous
GRANT ALL ON <table> TO authenticated; 
REVOKE <privileges> ON <table> FROM authenticated;
-- run command from first SQL
```

If anyone can share a better way of doing this, please feel free to share since this method is not quite clean.

**Reply by tanoabeleyra** [reply]
Thanks, this is highly needed. At the very minimum the `id` column shouldn't be writable.
I wrote the following function to make some columns read-only:

```sql
CREATE FUNCTION read_only_columns(table_name TEXT, ro_columns TEXT, roles TEXT DEFAULT 'authenticated')
RETURNS VOID
AS $$
DECLARE
  all_columns TEXT;
BEGIN
  SELECT array_to_string(array_agg(attname), ', ')
  INTO all_columns
  FROM pg_attribute
  WHERE attrelid = table_name::regclass AND attnum > 0 AND atttypid != 0;

  EXECUTE format('REVOKE INSERT, UPDATE ON %s FROM %s;', table_name, roles);
  EXECUTE format('GRANT INSERT(%1$s), UPDATE(%1$s) ON %2$s TO %3$s;', all_columns, table_name, roles);
  IF LENGTH(ro_columns) > 0 THEN
    EXECUTE format('REVOKE INSERT(%1$s), UPDATE(%1$s) ON %2$s FROM %3$s;', ro_columns, table_name, roles);
  END IF;
END;
$$ LANGUAGE plpgsql;
```

**Reply by GaryAustin1** [reply]
You can also use a trigger function on update to check for permission and if not allowed, then set new.col = old.col.
